### ISSUE TITLE
New principle: How to attach meta-information to a document

### ISSUE THREAD

**Issue by martinthomson** [w3ctag/design-principles#506]
There are several ways to attach meta-information to documents.  There are some trade-offs to understand for each.

* `data-*` attributes (which make information available to script and somewhat to styling)
* new custom elements (which make information available for presentation and styling, but can also be hidden)
* meta tags in the header
* link relations in the header
* HTTP header (or footer) fields

It might be good to list these out, point out that it is worth seeing if your use case is already addressed by a pre-existing thing for one of these other items, and go through some pros and cons of each.

Related to #213, and which might supersede #213.

**Comment by hober** [comment]
Not just documents, but also origins and sites. That is, this principle should also call out when to or when not to use `/.well-known/` URLs, when or when not to add stuff to Web Manifest, and the like.

**Comment by annevk** [comment]
Is the audience for this specification authors or web developers? As the former should never touch `data-*` or custom elements.

**Comment by bkardell** [comment]
I guess there is also `<script type="application/ld+json">` which is quite common and (I believe) not covered in the initial bullets of ways its been done.

**Comment by csarven** [comment]
I agree with the above that that it is important to understand who it is for. And, I think it may be more interesting or useful to encourage/discourage the use of "meta" information. Encourage more towards human-visible/readable.

---

That said, the `cite` attribute and `input type=hidden` are other examples used for the purpose of "meta".

**Comment by martinthomson** [comment]
In looking at this more closely, it seems like this is something more of a project than originally anticipated.

As others have noted, the question of audience is important.  We could say simple things like "this means of metadata attachment is for authors, not specification writers", which would cover HTML microsyntax and `data-*` attributes.  That's helpful to a degree, but the edges of even that are unclear to me.  And this is because the concept of metadata is a little bit confused.

I say confused, because metadata is just data.  What seems to matter is the audience (and maybe also the speaker).  For example, one line in HTML is between consumption by browsers as opposed to consumption by other things.  But there are many things that might act on HTML beyond the browser: accessibility technology, crawlers and search engines, etc...  Often, the entity consuming the thing is the JS running on the page itself.  Each picks up different things, uses them for different purposes, and we don't have a comprehensive understanding of the spectrum of uses.  Nor should we seek to do that.

Of course, there is potentially a far more pragmatic set of questions we could answer here.  We could say some basic things that would be helpful.  I've the start of text on `.well-known` that describes why you might use that (and why not).  The same for saying that `data-*` is off-limits for specification authors (this is already in its definition, so maybe we don't need to).  My main reservation there is that piecemeal advice here seems unsatisfactory and I can virtually guarantee that anything in this space will miss important things (I was not familiar with the use of `<script type="...json">`, for instance).
